neutral_generation/smart_convert.py

```diff
@@ -0,0 +1,284 @@
+from constants import *
+
+nlp = spacy.load("en_core_web_sm")
+
+# Load pre-trained model (weights)
+model = OpenAIGPTLMHeadModel.from_pretrained('openai-gpt')
+model.eval()
+# Load pre-trained model tokenizer (vocabulary)
+tokenizer = OpenAIGPTTokenizer.from_pretrained('openai-gpt')
+
+
+def convert(s):
+    # input: SNAPE sentence (meaning 1 entity and 1 gender)
+    # output: sentence in gender-neutral form
+
+    # pluralize verbs: change all verbs in third-person singular to third-person plural
+    auxiliaries = identify_verbs_and_auxiliaries(s)
+    s = pluralize_verbs(s, auxiliaries)
+
+    # use regex to replace pronouns and gendered terms that have a clear mapping
+    SIMPLE_REPLACE = EASY_PRONOUNS
+    SIMPLE_REPLACE.update(GENDERED_TERMS)
+    for p, r in SIMPLE_REPLACE.items():
+        s = regex_token_replace(s, p, r)
+
+    # use a LM to break ties for pronouns when there is a one-to-many mapping
+    for p, choices in NON_FUNCTION_PRONOUNS.items():
+        s = smart_pronoun_replace(s, p, choices)
+
+    return s
+
+
+def regex_token_replace(s, old, new):
+    # replace occurrences of a token in a string with a new token
+    # TODO: can look into redoing replace_map with SpaCy (they save capitalization as well)
+    replace_map = [[old, new], [capitalize(old), capitalize(new)], [old.upper(), new.upper()]]
+
+    for j in range(len(replace_map)):
+        pattern = re.compile(r'\b{}\b'.format(replace_map[j][0]))
+        # TODO: add an exception to regex replacement when an apostrophe follows the pronoun
+        # e.g. He'shan is a location, but with this rule it would become They'shan
+        s = re.sub(pattern, replace_map[j][1], s)
+
+    return s
+
+
+def capitalize(word):
+    return word[0].upper() + word[1:]
+
+
+def smart_pronoun_replace(s, p, choices):
+    # generate all choices, then choose best option using LM (one with lowest perplexity)
+    # TODO: currently uses OpenAI GPT, can try other LMs to see how performance changes
+    sentence_scores = dict()
+    for choice in choices:
+        new_sentence = regex_token_replace(s, p, new=choice)
+        if s != new_sentence:
+            new_score = score(new_sentence)
+            sentence_scores[new_sentence] = new_score
+
+    if not sentence_scores:   # source pronoun not found in sentence, meaning there are no choices to choose from
+        return s
+
+    print(sentence_scores)
+
+    return min(sentence_scores, key=sentence_scores.get)    # return sentence with lowest score (perplexity)
+
+
+def score(sentence):
+    tokenize_input = tokenizer.tokenize(sentence)
+    tensor_input = torch.tensor([tokenizer.convert_tokens_to_ids(tokenize_input)])
+    loss = model(tensor_input, lm_labels=tensor_input)
+    return math.exp(loss) / len(tokenize_input)     # normalize perplexity by number of tokens
+
+
+def identify_verbs_and_auxiliaries(s):
+    """
+    :param s: input sentence
```

Comment:
Does this have any normalization applied? I'm thinking lowercasing since the SUBJECT_PRONOUNS are "he" and "she"

Reply:
I believe I lowercase when checking for potential_subject, but I will do some testing to make sure that this captures our expected cases. Will also make this clearer in the code.